### DISCUSSION TITLE
ConstantBoundsAux and Constant/Monomial Nonlinear Variable

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#23443] [Q&A General] [answered]
Hi:

I am using an rDG scheme where my variables are constant/monomial.  I need to limit the range of one nonlinear variable (void fraction).  The ConstantBoundsAux seems perfect for this, but...

It appears that it only works if the nonlinear variable is nodal.  Is there a way to do this for an elemental variable.

Thanks,
 Joe Kelly

**Comment by GiudGiud** [comment]
@joshuahansel

**Reply by hugary1995** [reply]
It should be a straight-forward fix. You just need to generalize the BoundsAuxBase to take care of the elemental variable case. For example at https://github.com/idaholab/moose/blob/next/framework/src/auxkernels/BoundsAuxBase.C#L48-L49
I can walk you through this if any of you want to take a crack at it.

**Reply by joe61vette** [reply]
Wow, that was fast!  And yes, I am afraid that I'll need some help in modifying computeValue().  Thanks.

**Reply by hugary1995** [reply]
When I get off work today, I can give you the APIs to look at. Otherwise @GiudGiud may point you to those before I do. This bounds magic essentially boils down to figuring out the corresponding dof number of the nonlinear variable `bounded_variable` at the given node/element.

**Reply by joe61vette** [reply]
That would be great.  Thanks again.

**Reply by GiudGiud** [reply]
I can give it a shot this afternoon

**Reply by hugary1995** [reply]
Thanks @GiudGiud

**Reply by joe61vette** [reply]
Thanks @GiudGiud

**Reply by GiudGiud** [reply, marked as answer]
Should be good with https://github.com/idaholab/moose/pull/23449

@joshuahansel if you have time to check this out on RDG problems, could help determine if this is any help
(dont forget the extra petsc flag)

**Reply by joshuahansel** [reply]
@joe61vette Were you running into some cases that produced negative void fractions, or were you just being proactive? I don't recall fighting with negative values (just phase disappearance) since moving from CG to rDG, so it might be difficult for me to test this out (at least with the systems/formulations I work with).

**Reply by joe61vette** [reply]
Unfortunately, I can get negative void fractions and it is hard to recover from them.  I have "protected" my kernels and constitutive models by using the void fraction as a material property that is limited between 0 and 1.  But the nonlinear variable can still go negative when I am essentially solving a single-phase problem with the two fluid eqns.  Note, this is a simple air/water test case so no interfacial heat/mass transfer.

I first tried using the BoundingValueElementDamper.  But this quickly went to the minimum damping factor and stopped the problem.  I will give #23449 a try later today (THANKS @GiudGiud!)

**Reply by joe61vette** [reply]
To visualize how this has trouble recovering, let the volume fraction of one phase be negative.  Then, a positive velocity would convect mass INTO the cell rather than out of it.

**Reply by joe61vette** [reply]
#23449 works perfectly for my case!  Thanks Guillaume.  Now, on to the next problem....

**Comment by joshuahansel** [comment]
Hi Joe,

I'm assuming that you've already looked at https://mooseframework.inl.gov/source/auxkernels/ConstantBoundsAux.html, but I just wanted to make sure you saw this part:

> the user has to specify the PETSc options -snes_type vinewtonssls or -snes_type vinewtonrsls

I don't have any experience with PETSc's variational inequalities solvers, so I don't know how restrictive this is.

In RELAP-7 and Sockeye, we've come across the issue of bounding volume fractions, and we're using rDG/FV. I've never tried `ConstantBoundsAux` (I didn't know it existed!). I think for us, the issue was less about negativity of a volume fraction and more about the calculation of fluid properties of the absent/near-absent phase, since the thermodynamic state at which those properties are evaluated is very sensitive at those low volume fractions, and it's easy to go out of the range of validity in the fluid properties packages. I never found a great solution to this - I've tried many and gotten strategies that work well on some problems not work on others. I wish I had more time to explore this topic. Some general principles I tried to follow are:

- Start with a spatial discretization/stabilization that theoretically cannot get unphysical values. Of course, for a nonlinear solve, iterates can still be out of physical bounds, so see next bullet.
- Make sure the nonlinear residual calculation can handle unphysical values. You can bound the solution variable locally, for instance. This may lead to convergence difficulties due to a discontinuity in the nonlinear residual, so you can attempt to smooth this out with gradual transitions in the quantity/residual.

Even then, you can probably still end up with negative values... In the CG days of RELAP-7, we did this "volume fraction re-mapper" thing, that had mixed success, where we made our solution variable the unbounded void fraction (say, `alpha_unbounded`), that has a unique mapping to the bounded volume fraction (say, `alpha`), which is then what gets used in the residuals and such. So you could safely converge to a negative value of `alpha_unbounded`. Basically, you have `alpha_unbounded = alpha` for values that are larger than some small number `epsilon` and then you have a smooth mapping such that the limit of `alpha(alpha_unbounded)` as `alpha_unbounded` goes to infinity, is zero. I suspect that this approach brings new issues in this limit, but I don't know them well.

I'm sorry for rambling - it probably wasn't that helpful, since we never were completely successful.

Anyway, I'm interested to see how well `ConstantBoundsAux` works for you!

-Josh

**Reply by joe61vette** [reply]
Thanks Josh.  I figured that between Sockeye and RELAP-7 someone had already solved this.  If I can't figure out how to generalize ConstantBoundsAux to handle elemental variables, then I may have to do something like the mapping approach.  For some of my kernels, like the pressure gradient, I need a positive non-zero value of the phase volume fraction.  I'll let you know how this goes.

 Joe

**Comment by lindsayad** [comment]
Note that in the worst case you can take a nonlinear iteration per constrained dof to resolve active/inactive sets. As @joshuahansel has mentioned the best case scenario is to use a discretization scheme that is ensured to preserve physicality ... but even then TVD schemes have non-smooth derivatives which are themselves problematic for Newton
